### PR TITLE
Fix crash during daemon-reexec with systemd-oomd running

### PR DESCRIPTION
**Pull request by keszybz**
This fixes the crash. There's still some issue with oomd exiting with an error during the reexec, so further work will be needed.

### CODE REVIEW

**Comment by mrc0mmand** [comment]
I suspect packit doesn't fetch the `merge` GH ref, so you'll have to rebase this branch to get Packit green.

File: src/shared/varlink.c [hunk 1]
```diff
@@ -1219,17 +1220,23 @@ int varlink_close(Varlink *v) {
...
 
 Varlink* varlink_flush_close_unref(Varlink *v) {
         if (!v)
                 return NULL;
-
-        (void) varlink_flush(v);
-        (void) varlink_close(v);
+                                   /* n_ref >= 1 */
+        varlink_ref(v);            /* n_ref >= 2 */
+        varlink_flush(v);          /* n_ref >= 1 */
```
Comment (yuwata) [inline src/shared/varlink.c]: `varlink_flush()` does not decrease the reference counter. So the `/* n_ref >= 1 */` comment should moved one line below.

Maybe, more simply, it is better the following:
```
(void) varlink_flush(v);
return varlink_close_unref(v);
```

Comment (keszybz) [reply src/shared/varlink.c]: It *currently* doesn't. I included `varlink_flush` under protection for that reason. But yeah, maybe that's too much. I'll update code as suggested.

File: src/shared/varlink.c [hunk 2]
```diff
@@ -1219,18 +1220,19 @@ int varlink_close(Varlink *v) {
 Varlink* varlink_close_unref(Varlink *v) {
         if (!v)
                 return NULL;
-
-        (void) varlink_close(v);
+                                   /* n_ref >= 1 */
+        varlink_ref(v);            /* n_ref >= 2 */
+        varlink_close(v);          /* n_ref >= 1 */
+        if (v->n_ref > 1)
+                v->n_ref--;        /* n_ref >= 1 */
```
Comment (poettering) [inline src/shared/varlink.c:1242]: i don't get this? can you elaborate? now you sometimes drop a reference and sometimes you don't? that's looks very wrong to me?

Comment (keszybz) [reply src/shared/varlink.c:1242]: The sequence of events is:
when we enter `varlink_close_unref`, n_ref==1. We call varlink_close() which drops the last reference from a callback, so n_ref==0 and the object is deallocated. We then call varlink_unref() ourselves, and crash on access to freed memory.

By holding the reference to the object ourselves, we know the object is not deallocated. Essentially, we're saying: we take a reference and will drop it, unless the varlink_close() call we call happens to drop the next-to-last reference, in which case we'll just gracefully deallocate the object before returning.

We could say that freeing the object from the callback is not allowed, but that would be rather unpleasant.

Comment (poettering) [reply src/shared/varlink.c:1242]: Hmm, but how can n_ref==1 even be when entering varlink_close_unref() ?

**Comment by keszybz** [comment]
I added a big comment in the function explaining why we sometimes don't drop the reference, and I expanded the commit message to include the backtrace. No non-comment changes, except that I rebased, which should fix the packit failure.

**Comment by keszybz** [comment]
`TEST-53-ISSUE-16347:                FAIL     ( 33 s)` on amd64 looks unrelated. It passed in the previous iteration with no changes.

**Comment by keszybz** [comment]
So... this is a tough one. Nobody wants to review it, but it fixes a very significant issue and must be in -rc3. I'll merge it without review.

**Comment by bluca** [comment]
> So... this is a tough one. Nobody wants to review it, but it fixes a very significant issue and must be in -rc3. I'll merge it without review.

for what it's worth I did have a look when adding the labels, and while the changes themselves look fine, I cannot claim to follow and fully grok the ref/unref model, so didn't feel confident enough to add a green label on my own :-)

**Pull request merged**